apps/hash-graph/lib/graph/src/store/postgres/knowledge/entity.rs

```diff
@@ -40,251 +39,267 @@ impl<C: AsClient> PostgresStore<C> {
-            let Some(intersected_temporal_axes) = temporal_axes.intersect_variable_interval(variable_interval) else {
-                // `traverse_entity` is called with the returned entities from `read` with
-                // `temporal_axes`. This implies, that the version interval of `entity` overlaps
-                // with `temporal_axes`. `variable_interval` returns `None` if there are
-                // no overlapping points, so this should never happen.
-                unreachable!("the version interval of the entity does not overlap with the variable axis's time interval");
-            };
-
-            if current_resolve_depths.is_of_type.outgoing > 0 {
-                let entity_type_id =
-                    OntologyTypeVertexId::from(entity.metadata.entity_type_id().clone());
-                subgraph.edges.insert(Edge::KnowledgeGraph {
-                    vertex_id: entity_vertex_id,
-                    outward_edge: KnowledgeGraphOutwardEdge::ToOntology(OutwardEdge {
-                        kind: SharedEdgeKind::IsOfType,
-                        reversed: false,
-                        right_endpoint: entity_type_id.clone(),
-                    }),
-                });
-
-                self.traverse_entity_type(
-                    &entity_type_id,
-                    traversal_context,
-                    subgraph,
-                    GraphResolveDepths {
-                        is_of_type: OutgoingEdgeResolveDepth {
-                            outgoing: current_resolve_depths.is_of_type.outgoing - 1,
-                            ..current_resolve_depths.is_of_type
-                        },
-                        ..current_resolve_depths
-                    },
-                    intersected_temporal_axes.clone(),
-                )
-                .await?;
-            }
+        graph_resolve_depths: GraphResolveDepths,
+        traversal_context: &mut TraversalContext,
+        subgraph: &mut Subgraph,
+    ) -> Result<(), QueryError> {
+        let time_axis = temporal_axes.variable_time_axis();
 
-            if current_resolve_depths.has_left_entity.incoming > 0 {
-                for outgoing_link_entity in <Self as Read<Entity>>::read(
-                    self,
-                    &Filter::for_outgoing_link_by_source_entity_id(entity_vertex_id.base_id),
-                    &intersected_temporal_axes,
-                )
-                .await?
-                {
-                    let link_entity_interval = outgoing_link_entity
-                        .metadata
-                        .temporal_versioning()
-                        .variable_time_interval(time_axis);
+        let mut queue = entity_vertex_ids
+            .into_iter()
+            .map(|id| (id, graph_resolve_depths, temporal_axes.clone()))
+            .collect::<Vec<_>>();
+
+        while !queue.is_empty() {
+            for (entity_vertex_id, graph_resolve_depths, temporal_axes) in mem::take(&mut queue) {
```

Comment:
I actively decided to shadow the parameters here as while developing I accidentally used the initial parameters. The initial parameters are only for creating the initial queue and should never be used in the function again.

Reply:
That's the exact reason why I wouldn't want us to do this though. Adding `initial_` something makes it pretty clear if we're using the initial ones.

Whereas here I was confused while reading it which we were using, and while I appreciate it's not currently `mut`, a _potential_ change we could make would allow us to mutate it in place to decrement the depth, and then we could hit a really unpleasant bug to find.

I hear your reasoning, so I won't _ask_ you to change it, now that I've explained my concern (and also the confusion I had while reading) I'll leave it to your judgement which risk is higher

Reply:
I hear your concern as well. I'd prefer to keep it that way for now but will keep that concern in mind. I guess it's possible to get the best of both worlds when not passing around a tuple of things in the queue (which is a source of confusion here as well), but abstracting it in a struct.